woocommerce/qit-cli
# End-to-end tests for the QIT

This PR adds end-to-end tests for the Quality Insights Toolkit.

The E2E tests for the QIT does not use a browser, because the SUT is not a website, but the correct behavior of QIT as a whole.

### This PR:
- Runs a Security test against a given dev build with some security errors we expect to be flagged
- Asserts that the result of the test is as expected, by comparing a JSON snapshot of the results.
- Does the same for Activation tests.

Additionally, this was changed as requirements for the above to be possible:
- If the `--json` option is passed, wraps all output of the CLI and prevent any output that is not JSON from being printed.
- If the `--ignore-failure` option is passed when `--wait` is in use, always exits with status code 0, even if the test failed. (Used for scripting when we don't want a chain of command to be interrupted if the result of a test is a failure)

### Testing instructions
We can only test the GitHub workflows once we merge this PR, so these testing instructions will mimick what the GitHub workflow do:

- Checkout this branch
- Add new security violations that the Security tests should flag to the file `_tests/security/automatewoo/automatewoo.php`
- Run `make regenerate_security_snapshot`
- Assert that the snapshot now includes the violations that you introduced

Repeat the steps for Activation Tests _(PS: If multiple fatal errors are triggered in Activation Tests, only the first fatal is logged, because PHP execution is halted at runtime)_.

After we merge this branch, we can do a second round of testing, opening another PR and introducing some changes that we expect to be flagged, and manually triggering the workflow by going to Actions -> [Workflow Name] -> Run [branch-name]
